demo/src/components/logic.ts

```diff
@@ -1,32 +1,41 @@
 import { LogicFunction, ChessPiece } from './piece';
 
 export const pawnLogic: LogicFunction = (pawn, lookup) => {
-    const forward = (y: number, n: number) =>
-        pawn.color === 'white' ? y - n : y + n;
+    const forward = (n: number) =>
+        pawn.color === 'white' ? pawn.y - n : pawn.y + n;
 
     let result: [number, number][] = [];
-    if (lookup(pawn.x, forward(pawn.y, 1)) === undefined) {
-        result.push([pawn.x, forward(pawn.y, 1)]);
-    }
-    if (!pawn.wasMoved && lookup(pawn.x, forward(pawn.y, 1)) === undefined && lookup(pawn.x, forward(pawn.y, 2)) === undefined) {
-        result.push([pawn.x, forward(pawn.y, 2)]);
-    }
-    for (const x of [pawn.x - 1, pawn.x + 1]) {
-        if (x >= 0 && x < 8) {
-            let enemy = lookup(x, forward(pawn.y, 1));
-            if (enemy !== undefined && enemy.color !== pawn.color) {
-                result.push([x, forward(pawn.y, 1)]);
+
+    if (onBoard(forward(1))) {
+        if (lookup(pawn.x, forward(1)) === undefined) {
+            result.push([pawn.x, forward(1)]);
+        }
+        for (const x of [pawn.x - 1, pawn.x + 1]) {
+            if (onBoard(x)) {
+                let enemy = lookup(x, forward(1));
+                if (enemy !== undefined && enemy.color !== pawn.color) {
+                    result.push([x, forward(1)]);
+                }
             }
         }
     }
+    if (
+        !pawn.wasMoved &&
+        onBoard(forward(2)) &&
```

Comment:
This check is actually not needed, because if the pawn was not moved we know we are on board

Reply:
Yeah, wasn't sure about it.

Relying on initial board configuration feels dirty. On the other hand, rules do not generalize trivially. Like, should it be `pawn.wasMoved`, or `pawn.color === 'white' && pawn.x === 2 || pawn.color === 'black' && pawn.x === 7`, or both?

I'll remove it.